mypy/build.py

```diff
@@ -715,46 +721,125 @@ def stats_summary(self) -> Mapping[str, object]:
         else:
             assert st.meta, "Module must be either parsed or cached"
-            meta_snapshot[id] = st.meta.hash
+            hash = st.meta.hash
+        meta_snapshot[id] = hash
 
-    if not metastore.write(proto_meta, json.dumps(meta_snapshot)):
-        manager.log("Error writing protocol meta JSON file {}".format(proto_cache))
-        error = True
-    listed_proto_deps = {k: list(v) for (k, v) in proto_deps.items()}
-    if not metastore.write(proto_cache, json.dumps(listed_proto_deps)):
-        manager.log("Error writing protocol deps JSON file {}".format(proto_cache))
+    meta = {'snapshot': meta_snapshot, 'deps_meta': fg_deps_meta}
+
+    if not metastore.write(meta_path, json.dumps(meta)):
+        manager.log("Error writing fine-grained deps meta JSON file {}".format(meta_path))
         error = True
+
     if error:
         manager.errors.set_file(_cache_dir_prefix(manager), None)
-        manager.errors.report(0, 0, "Error writing protocol dependencies cache",
+        manager.errors.report(0, 0, "Error writing fine-grained dependencies cache",
                               blocker=True)
 
 
+def invert_deps(
+        deps: Dict[str, Set[str]],
+        graph: Graph) -> Tuple[Dict[str, Dict[str, Set[str]]], Dict[str, Set[str]]]:
+    from mypy.server.target import module_prefix, trigger_to_target
+
+    rdeps = {id: {} for id in graph}  # type: Dict[str, Dict[str, Set[str]]]
+    extra_deps = {}  # type: Dict[str, Set[str]]
+    for trigger, targets in deps.items():
+        module = module_prefix(graph, trigger_to_target(trigger))
+        if module and graph[module].tree:
+            mod_rdeps = rdeps.setdefault(module, {})
+            mod_rdeps.setdefault(trigger, set()).update(targets)
+        else:
+            extra_deps.setdefault(trigger, set()).update(targets)
+
+    return (rdeps, extra_deps)
+
+
+def process_deps(proto_deps: Dict[str, Set[str]],
+                 manager: BuildManager,
+                 graph: Graph) -> Tuple[Dict[str, Dict[str, Set[str]]],
+                                        Dict[str, Set[str]]]:
+    """Process fine-grained dependenecies into a form suitable for serializing.
+
+    Returns an (rdeps, extra_deps) pair, where rdeps maps from module ids to
+    all dependencies on that module, and extra_deps contains dependencies that
+    weren't associated with any module.
+    """
+
+    deps = manager.load_fine_grained_deps('@extra')
+    things = [st.compute_fine_grained_deps() for st in graph.values() if st.tree] + [proto_deps]
+    for st_deps in things:
+        for trigger, targets in st_deps.items():
+            deps.setdefault(trigger, set()).update(targets)
+
+    # If we are operating in non-incremental mode, properly split up
+    # dependencies between all the files. If this is an incremental update,
+    # though, we write all the dependencies into the "extra_deps" file so we
+    # don't need to reload dependency files to update them.
```

Comment:
Discuss what goes into `extra_deps` here. As far as I understand it, it doesn't cover all fine-grained deps -- only working set that was processed in the build? If everything would go there, it seems like this would make incremental cache generation somewhat inefficient. Anyway, if the extra deps contain a lot of stuff, will this make the optimization in this PR less useful? Would it sometimes be more efficient to split the dependencies in an incremental build even if we'd need to read all fine-grained deps?

Reply:
I was shoving all of the new stuff into extra_deps because it seemed simpler and unlikely to be that expensive on the daemon side. (Usually not a lot would be generated.) But having different behavior for `--incremental` and not is kind of bad, so I changed it to reload deps and split them in incremental runs.

I also got rid of `extra_deps` basically everywhere in favor of having it be `@root` in the normal map